lib/ecto/changeset.ex

```diff
@@ -2000,7 +2000,9 @@ defmodule Ecto.Changeset do
   @spec validate_inclusion(t, atom, Enum.t, Keyword.t) :: t
   def validate_inclusion(changeset, field, data, opts \\ []) do
     validate_change changeset, field, {:inclusion, data}, fn _, value ->
-      if value in data,
+      type = Map.get(changeset.types, field)
+
+      if Enum.any?(data, &Ecto.Type.equal?(type, value, &1)),
```

Comment:
This loses the optimization for ranges, but if you're thinking about ranges you probably should be using validate_number anyway?

Reply:
I was thinking about this too.
If you need numeric ranges, then validate_number provides better errors.

Reply:
Let's check the type. If the type is not an integer, float, or string, we use .equal?, otherwise `in` is fine. WDYT?

Reply:
Actually, we can move this to `Ecto.Type` and do something like this:

```elixir
  def include?(type, term, collection) do
    if fun = equal_fun(type) do
      Enum.any?(collection, &fun.(term, &1))
    else
      term in collection
    end
  end
```

It will remain efficient for base types and handle all equal types accordingly.

Reply:
Pushed an updated version